tools/tcptop.py

```diff
@@ -223,60 +240,60 @@ def pid_to_comm(pid):
         with open(loadavg) as stats:
             print("%-8s loadavg: %s" % (strftime("%H:%M:%S"), stats.read()))
 
-    # IPv4:  build dict of all seen keys
-    keys = ipv4_recv_bytes
-    for k, v in ipv4_send_bytes.items():
-        if k not in keys:
-            keys[k] = v
+    ipv4_send_bytes_dict = ipv4_table_to_dict(ipv4_send_bytes)
+    ipv4_recv_bytes_dict = ipv4_table_to_dict(ipv4_recv_bytes)
+
+    # IPv4: build dict of all seen keys
+    keys = ipv4_send_bytes_dict.copy()
+    keys.update(ipv4_recv_bytes_dict)
 
     if keys:
         print("%-6s %-12s %-21s %-21s %6s %6s" % ("PID", "COMM",
             "LADDR", "RADDR", "RX_KB", "TX_KB"))
 
     # output
-    for k, v in reversed(sorted(keys.items(), key=lambda keys: keys[1].value)):
+    for k, v in sorted(keys.items(), key=lambda kv: kv[1], reverse=True):
         send_kbytes = 0
-        if k in ipv4_send_bytes:
-            send_kbytes = int(ipv4_send_bytes[k].value / 1024)
+        if k in ipv4_send_bytes_dict:
+            send_kbytes = int(ipv4_send_bytes_dict[k] / 1024)
         recv_kbytes = 0
-        if k in ipv4_recv_bytes:
-            recv_kbytes = int(ipv4_recv_bytes[k].value / 1024)
+        if k in ipv4_recv_bytes_dict:
+            recv_kbytes = int(ipv4_recv_bytes_dict[k] / 1024)
 
         print("%-6d %-12.12s %-21s %-21s %6d %6d" % (k.pid,
             pid_to_comm(k.pid),
-            inet_ntop(AF_INET, pack("I", k.saddr)) + ":" + str(k.lport),
-            inet_ntop(AF_INET, pack("I", k.daddr)) + ":" + str(k.dport),
+            k.laddr + ":" + str(k.lport),
+            k.daddr + ":" + str(k.dport),
             recv_kbytes, send_kbytes))
 
     ipv4_send_bytes.clear()
```

Comment:
Let's now clear right after we read. The clearing is also an iteration through the tables and there are definitely race between clearing them from user space and BPF program still writing things to it.